Allow to prewarm the cache for searchable snapshot shards

tlrx opened the pull request:
This pull requests adds a way to prewarm the cache for searchable snapshot shard files. 

It relies on a new index setting named `index.store.snapshot.cache.load.eagerly` (defaults to `false`) that can be passed when mounting a snapshot as an index. This setting is detected during the pre-recovery step before the snapshot files are exposed to the other components of the system. The method `prewarmCache()` of the `SearchableSnapshotDirectory` instance is executed, which builds the list of all parts of snapshot files that needs to be prefetched in cache (excluding the files that are stored in metadata hash and the ones explicitly excluded by the `excluded_file_types` setting). 

Then parts are prefetched in cache in parallel using the `SNAPSHOT` thread pool. If a snapshot file is composed of multiple parts (or chunks) then the parts can potentially be downloaded and written in cache concurrently. The implementation relies on a new `prefetchPart()` method added to the  `CachedBlobContainerIndexInput` class. This method allows to fetch a complete part of a file (or the whole file if the snapshot file is composed of a single part) in order to write it in cache. This is possible because `CacheFile` has been modified to work with configurable cache range sizes depending on the `IOContext` the `IndexInput` has been opened with. 

When the `IndexInput` is opened using the specific `CACHE_WARMING_CONTEXT` context then the file is cached on disk using large ranges of bytes aligned on the beginning and the end of each part (or chunk) of the file. When using a different context then the fill is cached on disk using the normal cache range size defined through the `range_size` setting. This implementation allows to reuse the existing cache eviction mechanism if something goes wrong when reading or writing the part. It also simplifies the logic if the recovering shard is closed while prewarming the cache.

tlrx committed:
Add cache prewarming

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -115,8 +135,10 @@ public SearchableSnapshotDirectory(
         this.cacheDir = Objects.requireNonNull(cacheDir);
         this.closed = new AtomicBoolean(false);
         this.useCache = SNAPSHOT_CACHE_ENABLED_SETTING.get(indexSettings);
+        this.loadCacheEagerly = useCache ? SNAPSHOT_CACHE_LOAD_EAGERLY_SETTING.get(indexSettings) : false;
```
Better name suggestions are welcome

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -142,18 +164,21 @@ protected final boolean assertCurrentThreadMayLoadSnapshot() {
                     this.blobContainer = blobContainerSupplier.get();
                     this.snapshot = snapshotSupplier.get();
+                    if (loadCacheEagerly) {
+                        prewarmCache();
```
This methods blocks until the cache is fully prewarmed. It must be done before `loaded` is set to `true` so that other components of the system are not likely to trigger some caching on this directory files.

ywelsch replied:
Is that strictly necessary? I would prefer to initiate the prewarming here, but at the same time allow the shard routing to move to started state as quickly as possible.

tlrx replied:
This is not strictly necessary but a misunderstanding from my part. We discussed this and I updated the PR so that cache warming now runs concurrently with the recovery.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -290,17 +315,20 @@ public CacheFile getCacheFile(CacheKey cacheKey, long fileLength) throws Excepti
+    }
 
-        final BlobStoreIndexShardSnapshot.FileInfo fileInfo = fileInfo(name);
+    private IndexInput openInput(final BlobStoreIndexShardSnapshot.FileInfo fileInfo, final IOContext context) {
```
Splitting this method into two allows to open an `IndexInput` even if the snapshot is not marked as `loaded` yet

tlrx replied:
Indeed - I pushed c8a1c6bf7e480d4961fe1e4a6d7fb8ab01e85534 to remove the method.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -331,12 +359,133 @@ public String toString() {
+                    );
+                    final long startTimeInNanos = statsCurrentTimeNanosSupplier.getAsLong();
+                    try {
+                        final IndexInput input = openInput(file, CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
```
This method uses an `IndexInput` with a specific `IOContext` to prewarm the cache for the given Lucene file. The `IndexInput` will be cloned for each part to write in cache later and closed once all parts are processed.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -331,12 +359,133 @@ public String toString() {
+                        } catch (Exception e) {
+                            if (e instanceof AlreadyClosedException
+                                || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
+                                continue;
```
This can happen in case of cache evictions or if the shard is closing

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -61,12 +61,11 @@ protected void closeInternal() {
     private volatile FileChannel channel;
 
-    public CacheFile(String description, long length, Path file, int rangeSize) {
+    public CacheFile(String description, long length, Path file) {
```
We included the rangeSize in the CacheFile to compute the range to fetch given a specific position, but we were never asserting that the fetched ranges really matched the size.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -144,6 +208,7 @@ private void writeCacheFile(FileChannel fc, long start, long end) throws IOExcep
         final long length = end - start;
         final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, length))];
         logger.trace(() -> new ParameterizedMessage("writing range [{}-{}] to cache file [{}]", start, end, cacheFileReference));
+        assert assertRangeOfBytesAlignment(start, end);
```
This asserts the size of the ranges written in cache, depending of the IOContext.

tlrx replied:
We can't really assert the size of ranges now warming runs concurrently. This has been removed.

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -131,6 +170,31 @@ protected void readInternal(final byte[] buffer, final int offset, final int len
+        final Tuple<Long, Long> range = computeRange(IntStream.range(0, part).mapToLong(fileInfo::partBytes).sum());
+        try {
+            final CacheFile cacheFile = getCacheFileSafe();
+            try (ReleasableLock ignored = cacheFile.fileLock()) {
```
I'm wondering if this is a source of contention; I'll investigate further.

ywelsch commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -331,12 +359,133 @@ public String toString() {
+                @Override
+                protected void doRun() throws Exception {
+                    CheckedRunnable<Exception> loader;
+                    while (isOpen && (loader = queue.poll(0L, TimeUnit.MILLISECONDS)) != null) {
```
I see that this is taking the same approach as we use for uploading snapshot files (BlobStoreRepository). I would prefer not to hold onto workers for such a long time, as it can block the snapshot thread pool for a long time (cc: @original-brownbear).
In both cases (also the one BlobStoreRepository), I would prefer for the worker to process one file, then enqueue another task to the thread pool to pick up the next piece of work. This allows other operations to make progress as well, instead of waiting for a long time in the snapshot queue.

tlrx replied:
I agree, that makes sense Yannick.

original-brownbear replied:
I think for the uploading side the downside of

>  This allows other operations to make progress as well

is that it causes the index commits to be held on for a suboptimally long time. That's why the approach of fully monopolizing the pool was consciously chosen for uploads there.

ywelsch replied:
That's something to be better controlled at the `SnapshotShardsService` level then, though. It could limit the number of shards to be snapshotted concurrently by lazily enqueuing there.

tlrx commented:
Please hold on reviews - Yannick and I discussed several points on this PR yesterday and I'll addresse them.

tlrx committed:
Merge branch 'master' into load-cache-eagerly

tlrx committed:
apply feedback

tlrx commented:
I've updated this PR so that the cache warming is not blocking anymore and now runs concurrently with the shard recovery. Allowing concurrent reads of different chunk sizes required to remove assertions on the number and the length of gaps to be written in cache (which I think is OK as David planned to improve this). I also introduced a dedicated thread pool for cache warming as suggested by Yannick. This thread pool is sized larger than the default snapshot thread pool.

I've quickly ran some benchmarks and compared the results to regular full restores. Depending of the snapshot to restore this change runs from 10% to 50% faster than a regular restore. This makes sense now contention have been reduced in #55662 and custom thread pool is used for warming.

tlrx commented:
[ML related](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/22693/) failure

@elasticmachine run elasticsearch-ci/2

ywelsch commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -331,12 +353,85 @@ public String toString() {
+                    final CountDown countDown = new CountDown(Math.toIntExact(numberOfParts));
+                    for (long p = 0; p < numberOfParts; p++) {
+                        final int part = Math.toIntExact(p);
+                        executor.execute(new AbstractRunnable() {
```
This will queue up all chunks at once on the thread pool. To make this fairer among multiple recoveries, I would prefer something like https://github.com/elastic/elasticsearch/compare/master...ywelsch:para-restores?expand=1

Can be a follow-up though if you prefer.

tlrx replied:
Yes I still have this in mind - I'd prefer to do it in a follow up PR if possible.

DaveCTurner replied:
Could you transfer that to a `// TODO` comment, just so we are sure not to lose it?

tlrx replied:
TODO added in a635c7f064604d74bec4c520580846042aee6310

tlrx replied:
I opened https://github.com/elastic/elasticsearch/pull/55793 for this.

tlrx commented:
@DaveCTurner sorry for the delay. This is ready for review.

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -249,41 +248,36 @@ private void ensureOpen() {
 
     CompletableFuture<Integer> fetchRange(
-        long position,
+        Tuple<Long, Long> range,
```
A `Tuple` argument is a bit strange, can we pass in the two `long`s directly instead?

tlrx replied:
Sure. I pushed 4dc76b83c10f3881a7783896dd3610bc7baae522

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -249,41 +248,36 @@ private void ensureOpen() {
         try {
-            if (position < 0 || position > tracker.getLength()) {
-                throw new IllegalArgumentException("Wrong read position [" + position + "]");
+            if (range.v1() < 0 || range.v1() > tracker.getLength() || range.v2() < 0 || range.v2() > tracker.getLength()) {
```
Also invalid if the start of the range is after the end?

tlrx replied:
Right, I pushed 4dc76b83c10f3881a7783896dd3610bc7baae522

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:
```diff
@@ -249,41 +248,36 @@ private void ensureOpen() {
                 )
             );
 
             if (gaps.size() > 0) {
```
Don't really need this guard any more.

tlrx replied:
Removed in e9831a3c1c5f2035586b08bf390667a149313359

DaveCTurner reviewed (commented):
Cunning use of an `IOContext`, makes sense to me. I left some small comments but nothing major.

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:
```diff
@@ -131,6 +172,31 @@ protected void readInternal(final byte[] buffer, final int offset, final int len
+        if (part >= fileInfo.numberOfParts()) {
+            throw new IllegalArgumentException("Unexpected part number [" + part + "]");
+        }
+        final Tuple<Long, Long> range = computeRange(IntStream.range(0, part).mapToLong(fileInfo::partBytes).sum());
```
Can we assert that the returned range is the whole part? I think that'd be useful documentation if nothing else.

tlrx replied:
Sure. I added e01bf710ffa6a00ab730fc3efe45b4204dadd763, let me know what you think.

tlrx committed:
remove unused openinput

tlrx committed:
add todo

tlrx committed:
rename setting

tlrx committed:
assert thread name

tlrx committed:
remove tuple

tlrx committed:
remove unnecessary condition

tlrx committed:
assert range

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java:
```diff
@@ -129,7 +129,7 @@ protected InputStream openSlice(long slice) throws IOException {
     }
 
-    protected final boolean assertCurrentThreadMayAccessBlobStore() {
+    protected boolean assertCurrentThreadMayAccessBlobStore() {
```
Ok. I pushed 0469c7b582ffa55b267d082c41c29221b3e6e876

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -99,6 +103,11 @@
         Setting.Property.IndexScope
     );
+    public static final Setting<Boolean> SNAPSHOT_CACHE_LOAD_EAGERLY_SETTING = Setting.boolSetting(
+        "index.store.snapshot.cache.load.eagerly",
```
Much better name, thanks. I pushed 4ee96afd25d50d941bfd43e0a4c558cfded45b54.

tlrx commented:
Thanks @DaveCTurner, I've applied your feedback. This is ready for another round.

DaveCTurner reviewed (approved):
Two more tiny comments regarding the threadpool, but otherwise LGTM. Great work @tlrx.

tlrx committed:
zero

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -252,4 +266,21 @@ public void onIndexModule(IndexModule indexModule) {
+    public static ExecutorBuilder<?> executorBuilder() {
+        return new ScalingExecutorBuilder(
+            SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME,
+            1,
```
Good point - I pushed baa8258283eb91a45d3fc433320bce0a8aaca191

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -252,4 +266,21 @@ public void onIndexModule(IndexModule indexModule) {
 
+    public List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
+        if (SEARCHABLE_SNAPSHOTS_FEATURE_ENABLED) {
+            return List.of(executorBuilder());
```
The prewarming is enabled or disabled on a per-index basis using a `IndexScope` setting. There is no mechanism to disable prewarming globally for the node. The threadpool is created globally at node construction time and I don't think we can prevent its creation as the node might be assigned later an index with cache prewarming enabled.

DaveCTurner replied:
D'oh yes good point I thought it was a global setting for some reason. Don't mind me ...

tlrx commented:
Thanks David
